pallets/ds-maps/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+decl_event!(
+    pub enum Event<T>
+    where
+        AccountId = <T as frame_system::Trait>::AccountId,
+        RootId = <T as Trait>::RootId,
+        AreaId = <T as Trait>::AreaId,
+        ZoneId = <T as Trait>::ZoneId,
+        {
+        // Event documentation should end with an array that provides descriptive names for event parameters.
+        /// New root box has been created [box number, who]
+        RootCreated(RootId, AccountId),
+        /// New zone added [root, area, zone number, who]
+        ZoneCreated(RootId, AreaId, ZoneId, AccountId),
+        /// area type changed [role, area, root, who]
+        AreaTypeChanged(u8, AreaId, RootId, AccountId),
+    }
+);
+
+// Errors inform users that something went wrong.
+// learn more https://substrate.dev/docs/en/knowledgebase/runtime/errors
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        /// Error names should be descriptive.
+        NoneValue,
+        /// Operation is not valid
+        InvalidAction,
+        /// Incorrect data provided
+        InvalidData,
+        /// Origin do not have sufficient privileges to perform the operation
+        NotAuthorized,
+        /// Account doesn't exist
+        NotExists,
+        /// Area is unavailable for operation
+        ForbiddenArea,
+        /// Trying to add zone in non-existing root
+        RootDoesNotExist,
+        /// Sizes are off bounds
+        BadDimesions
+        // add additional errors below
+    }
+}
+
+// Dispatchable functions allows users to interact with the pallet and invoke state changes.
+// These functions materialize as "extrinsics", which are often compared to transactions.
+// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        // Errors must be initialized if they are used by the pallet.
+        type Error = Error<T>;
+
+        // Events must be initialized if they are used by the pallet.
+        fn deposit_event() = default;
+
+        #[weight = <T as Trait>::WeightInfo::root_add()]
+        pub fn root_add(origin, 
+                        bounding_box: Box3D<Point3D<T::Coord>>,
+                        delta: T::Coord) -> dispatch::DispatchResult {
+            let who = ensure_signed(origin)?;
+            ensure!(<accounts::Module<T>>::account_is(&who, REGISTRAR_ROLE.into()), Error::<T>::NotAuthorized);
+            // Here more complex calculation for root dimensions and delta needed
+            let lat_dim = bounding_box.south_east.lat - bounding_box.north_west.lat;
+            ensure!(lat_dim.into() <= U9F23::from_num(1f64), Error::<T>::BadDimesions);
+            let lon_dim = bounding_box.south_east.lon- bounding_box.north_west.lon;
+            ensure!(lon_dim.into() <= U9F23::from_num(1f64), Error::<T>::BadDimesions);
+            ensure!(delta.into() <= U9F23::from_num(0.1f64), Error::<T>::InvalidData);
```

Comment:
Rather tight coupling to U9F23 type.
I'd say, it is inappropriate to mention U9F23 in this module at all.
T::Coord must be `From<f64>` and then you can compare `lat_dim < 1_f64.into()`.
Let's discuss via voice if unclear (this is rather broad topic, actually).

Reply:
Also, minimum dimension of 1x1x0.1 looks pretty arbitrary and unjustified. Why exactly these restrictions? Why not to parameterize at least?

Reply:
Maximum dimensions were recalculated and correct now. 
I proceeded from the fact that the dimensions should be described by u16. That is, a maximum of 65536 areas.